Runtime/ui/txt/layout.cs

```diff
@@ -1,218 +1,284 @@
+            float[] advances, int advanceOffset, TabStops tabStops) {
+            char startingChar = text[start];
+            float totalWidths = 0;
+            if (char.IsHighSurrogate(startingChar) || EmojiUtils.isSingleCharEmoji(startingChar)) {
+                float advance = style.fontSize + style.letterSpacing;
+                for (int i = 0; i < count; i++) {
+                    char ch = text[start + i];
+                    if (char.IsHighSurrogate(ch) || EmojiUtils.isSingleCharNonEmptyEmoji(ch)) {
+                        advances[i + advanceOffset] = advance;
+                        totalWidths += advance;
+                    }
+                    else {
+                        advances[i + advanceOffset] = 0;
+                    }
+                }
+            }
+            else {
+                Font font = FontManager.instance.getOrCreate(style.fontFamily, style.fontWeight, style.fontStyle).font;
+                // TODO: it is kind of a waste to require the entire string for this style, but SubString causes alloc
+                font.RequestCharactersInTextureSafe(text, style.UnityFontSize, style.UnityFontStyle);
+                for (int i = 0; i < count; i++) {
+                    char ch = text[start + i];
+                    if (ch == '\t') {
+                        advances[i + advanceOffset] = tabStops.nextTab(offset + totalWidths) - (offset + totalWidths);
+                    }
+                    else if (font.getGlyphInfo(ch, out var glyphInfo, style.UnityFontSize, style.UnityFontStyle)) {
+                        advances[i + advanceOffset] = glyphInfo.advance + style.letterSpacing;
+                    }
+                    else {
+                        advances[i + advanceOffset] = style.letterSpacing;
+                    }
+
+                    if (LayoutUtils.isWordSpace(ch)) {
+                        advances[i + advanceOffset] += style.wordSpacing;
+                    }
+
+                    totalWidths += advances[i + advanceOffset];
+                }
+            }
+
+            return totalWidths;
+        }
+
+        public static float doLayout(float offset, string text, int start, int count, TextStyle style,
+            float[] advances, float[] positions, TabStops tabStops, out UnityEngine.Rect bounds) {
+            float advance = 0;
+            Font font = FontManager.instance.getOrCreate(style.fontFamily, style.fontWeight, style.fontStyle).font;
+
+            char startingChar = text[start];
+            bounds = new UnityEngine.Rect();
+            if (char.IsHighSurrogate(startingChar) || EmojiUtils.isSingleCharEmoji(startingChar)) {
+                advance = _layoutEmoji(text, start, count, style, font, advances, positions, ref bounds);
+            }
+            else {
+                // According to the logic of Paragraph.layout, it is assured that all the characters are requested

```

Comment:
make sure the new logic is the same as in the original code?

Reply:
I tried to keep the logic intact, and fixed some issues that exists in the original code. I tested with all the samples in the project that contain texts and editable texts and they all look the same as before.